catalog/cataloger_merge.go

```diff
@@ -52,49 +54,46 @@ func (c *cataloger) Merge(ctx context.Context, repository, leftBranch, rightBran
 				Limit: -1,
 			},
 		}
-
-		scanner, err := NewDiffScanner(tx, params)
+		relation, err := getRefsRelationType(tx, params)
 		if err != nil {
 			return nil, err
 		}
-		if scanner.Relation == RelationTypeSame {
+		if relation == RelationTypeSame {
 			return nil, ErrSameBranchMergeNotSupported
 		}
 		nextCommitID, err := getNextCommitID(tx)
 		if err != nil {
 			return nil, err
 		}
-		// do the merge based on the relation
 		previousMaxCommitID, err := getLastCommitIDByBranchID(tx, rightID)
 		if err != nil {
 			return nil, err
 		}
-
-		differences := make(mergeBatchRecords, 0, MergeBatchSize)
+		mergeBatchChan := make(chan mergeBatchRecords, MergeBatchChanBuffer)
+		var workerExitFlag bool
+		go c.diffWorker(params, mergeBatchChan, &workerExitFlag)
```

Comment:
the worker needs to accept the call 'ctx' and check/honor it in case of cancelation.

Reply:
AFAIU, Ctx helps when it is combined with other wait states - e.g. wait on a channel and context. The diffWorker is very simple and has no select statement